wdl/qc_imputation.wdl

```diff
@@ -122,53 +156,423 @@ workflow qc_imputation {
+        $plink2_cmd --bfile merged --maf ${pca_maf} --hwe ${hwe_for_pca} --geno ${variant_missing_for_pca} --exclude range ${high_ld_regions} \
+            --indep-pairwise ${pca_ld} --out pruned
+        $plink2_cmd --bfile merged --mind ${sample_missing_for_pca} --pca --extract pruned.prune.in --out ${base}
+
+        echo -e "`date`\trun glm"
+        # logistic glm between dataset and panel
+        awk 'BEGIN{OFS="\t"} {print "0", $2, "1"}' panel_gt_isec.fam > pheno
+        awk 'BEGIN{OFS="\t"} {print "0", $2, "2"}' gt_panel_isec.fam >> pheno
+        $plink2_cmd --bfile merged --mind ${sample_missing_for_pca} --const-fid --glm firth-fallback --covar ${base}.eigenvec --pheno pheno --out ${base}
+
+        echo -e "`date`\tcompute frequencies"
+        # extract stats from the glm result and calculate af in dataset and panel
+        grep -E "TEST|ADD" ${base}.PHENO1.glm.logistic.hybrid > temp && mv temp ${base}.PHENO1.glm.logistic.hybrid
+        $plink2_cmd --bfile gt_panel_isec --freq --out ${base}
+        $plink2_cmd --bfile panel_gt_isec --freq --out ${base}_panel_freq
+        awk 'BEGIN{OFS="\t"} NR==1{print "CHR","SNP","REF","ALT","AF"} NR>1{print "chr"$1,$2,$3,$4,$5}' ${base}.afreq > ${base}.frq
+        awk 'BEGIN{OFS="\t"} NR==1{print "CHR","SNP","REF","ALT","AF"} NR>1{print "chr"$1,$2,$3,$4,$5}' ${base}_panel_freq.afreq > ${base}_panel_freq.frq
+        echo -e "`date`\tdone"
+
+    >>>
+
+    output {
+        File glm  = base + ".PHENO1.glm.logistic.hybrid"
+        File dataset_freq_panel_intersect = base + ".frq"
+        File panel_freq_dataset_intersect = base + "_panel_freq.frq"
+        File eigenvec = base + ".eigenvec"
+    }
+
+    runtime {
+        docker: "${docker}"
+        memory: ceil(size(panel_bed, "G") / 5) + " GB"
+        cpu: 4
+        disks: "local-disk 200 HDD"
+        preemptible: 2
+        zones: "europe-west1-b europe-west1-c europe-west1-d"
+    }
+}
+
+task missingness {
+
+    String base
+    File bed
+    File bim
+    File fam
+    Float variant_missing_n_batches
+    String docker
 
-    Array[File] vcfs
-    String outfile
     String dollar = "$"
-    String storage="local-disk 300 HDD"
-    Int cpus = 32
+
+    command <<<
+
+        set -euxo pipefail
+
+        mem=$((`free -m | grep -oP '\d+' | head -n 1`-100))
+        plink_cmd="plink --allow-extra-chr --keep-allele-order --memory $mem"
+
+        $plink_cmd --bfile ${sub(bed, "\\.bed", "")} --missing --out ${base}
+
+        awk -v batch=${base} 'BEGIN {OFS="\t"} \
+            NR==1 {
+                for(i=1;i<=NF;i++) { h[$i]=i }; \
+                if( !("F_MISS" in h && "SNP" in h) ) \
+                    {print "F_MISS and SNP columns expected in plink output." > "/dev/stderr"; exit 1; \
+                }  \
+            } \
+            NR>1 { \
+                if($h["F_MISS"]>${variant_missing_n_batches}) { \
+                    print $h["SNP"],"joint_qc_batch_high_missing",batch":"$h["F_MISS"]
```

Comment:
can you remove the "joint_qc" from the beginning of the reject reason and have the reason simply be "batch_high_missing".

Reply:
there seems to be batch missingness threshold twice
  "qc_imputation.missingness.variant_missing_n_batches": 0.04,
  "qc_imputation.batch_qc.variant_missing": 0.04
  Remove the other one and without reading further. Is this step needed at all if there is batch missingess step later?

Reply:
variant_missing_n_batches means if missingness is > X in N batches, remove from all batches

variant_missing means if missingness > X in a batch, remove from that batch

Reply:
joint_qc_ I'd rather keep so it's easy to afterward track which variants were excluded by joint qc

Reply:
> joint_qc_ I'd rather keep so it's easy to afterward track which variants were excluded by joint qc

ah yea my bad, for some reason was thinking that this was in batch qc but indeed not at all.

Reply:
> variant_missing_n_batches means if missingness is > X in N batches, remove from all batches
> 
> variant_missing means if missingness > X in a batch, remove from that batch

yep got that but just wondering if we need separate thresholds for those two as they are applied per batch anyway first and then in the other tested if we have missing in more than N.  Off course it adds flexibility if we ever need such a feature. Just a consideration to try to minimize the number of moving parts and simplify